### DISCUSSION TITLE
Flink CDC2.1.1 varchar类型主键大表数据同步速度很慢的问题

### DISCUSSION THREAD

**Discussion by a120610114** [apache/flink-cdc#801] [Q&A]
在使用FlinkCDC2.1.1 版本进行数据同步的时候我发现对于主键是varchar类型的大表数据（大约3000w左右的数据）数据同步非常慢，而且还经常报错。在查看日志的时候发现了这样的日志
`fb72562c-845f-11ea-88a6-b8599fe5d1ea:1-244504311, row=0, event=0} for split MySqlSnapshotSplit{tableId=wxqyh_learnonline.tb_qy_examination_exam_user_ref, splitId='wxqyh_learnonline.tb_qy_examination_exam_user_ref:2031', splitKeyType=[`id` VARCHAR(32) NOT NULL], splitStart=[b6547e5906a64ae6a5b062492f1b4b91], splitEnd=[b66b1de7513043379a327af9086c8f80], highWatermark=null}
2022-01-07 00:28:28,429 INFO  com.ververica.cdc.connectors.mysql.debezium.task.MySqlSnapshotSplitReadTask [] - Snapshot step 2 - Snapshotting data
2022-01-07 00:28:28,429 INFO  com.ververica.cdc.connectors.mysql.debezium.task.MySqlSnapshotSplitReadTask [] - Exporting data from split 'wxqyh_learnonline.tb_qy_examination_exam_user_ref:2031' of table wxqyh_learnonline.tb_qy_examination_exam_user_ref
2022-01-07 00:28:28,429 INFO  com.ververica.cdc.connectors.mysql.debezium.task.MySqlSnapshotSplitReadTask [] - For split 'wxqyh_learnonline.tb_qy_examination_exam_user_ref:2031' of table wxqyh_learnonline.tb_qy_examination_exam_user_ref using select statement: 'SELECT * FROM `wxqyh_learnonline`.`tb_qy_examination_exam_user_ref` WHERE id >= ? AND NOT (id = ?) AND id <= ?'
2022-01-07 00:28:31,217 INFO  com.ververica.cdc.connectors.mysql.debezium.task.MySqlSnapshotSplitReadTask [] - Finished exporting 8093 records for split 'wxqyh_learnonline.tb_qy_examination_exam_user_ref:2031', total duration '00:00:02.788'
2022-01-07 00:28:31,222 INFO  com.ververica.cdc.connectors.mysql.debezium.task.MySqlSnapshotSplitReadTask [] - Snapshot step 3 - Determining high watermark {ts_sec=0, file=mysql-bin.001188, pos=53630482, gtids=1935d4a6-2e71-11e9-9330-6c92bf5f0aed:450812989-484900678,
4ddb2c3f-6f05-11e9-8a9c-6c0b84d5a828:1-411556268,`
这让我很疑惑，查看源码也是一样的
![企业微信截图_16417826015078](https://user-images.githubusercontent.com/26463863/149500906-6d61b10e-3717-4c0d-9c49-f76b84e3e9f2.png)
![企业微信截图_16417826231717](https://user-images.githubusercontent.com/26463863/149500932-37f0e365-c14b-4824-b861-b8eba6d679db.png)
这是直接对varchar类型进行切分然后查询，当我用这个语句去数据库查询的时候发现非常慢
![企业微信截图_16421514818830](https://user-images.githubusercontent.com/26463863/149501085-424ac8ed-5f46-4eb7-a7b4-842c062b299f.png)
我想问下这种想象是正常的吗？设计的原理是什么？有没有什么办法可以提高这种varchar 大表数据的同步？

**Comment by a120610114** [comment]
@wuchong 大佬可以帮忙解答一下吗

**Reply by xiaoyu19950911** [reply]
非自增的主键每次分chunk都会去查一遍db，你的chunkSize看起来是默认大小，可以尝试一下调大chunkSize试试

**Reply by a120610114** [reply]
除了这个还有其他的办法吗？调整这个参数 debezium.min.row. count.to.stream.result 是否会有用呢？改为流式读取
![image](https://user-images.githubusercontent.com/26463863/149732358-431b1e77-c991-4fcd-871d-1266596092f0.png)

**Reply by xiaoyu19950911** [reply]
你最好先看下速度很慢耗时在哪，如果是耗时在切片的话，那这个参数的调整也没有意义，其次，分片同步是没有使用debzium原生的全量同步逻辑的，所以调整这个参数也不会生效的，分片同步控制全量部分的参数为fetchSize

**Reply by a120610114** [reply]
其实我比较担心的是 非自增的主键的分片，如果表的数量太大的话 是不是后面会影响性能？或者说后期的时候根本读不出来。因为我的任务在运行了几个小时后就莫名挂了，这种情况出现了几次。但是如果是自增型主键的表的话 数据量更大都没有问题

**Comment by wuchong** [comment]
现在varchar 类型慢，一般是卡在前面计算切片的时候（要把所有切片都分好，才开始每个切片的读取）。有一个优化思路是异步切片：
1. enumerator 通过异步线程去切表的切片。
2. readers 来取切片的时候，只要有一个切片就能开始读取，就不用等所有切片分好。
3. reader 在读数据的过程中，enumerator 仍在异步分切片。

**Reply by a120610114** [reply]
> 

这个Flink CDC目前具备这个功能吗？还是说后续的一个优化方向

**Reply by wuchong** [reply]
是后续的一个优化方向，可以开个issue。

**Reply by a120610114** [reply]
> 是后续的一个优化方向，可以开个issue。
想要再问一个问题，varchar 类型的chunk 切分机制是什么啊？我在测试环境的时候对测试数据库进行数据同步时我发现 chunk分出来的数据是很均匀的，chunkSize 设置为5w 那分出来就是5w。 当是当我将程序丢入生产环境的时候发现 chunk出来的数据就不均匀了，前面的chunk都是5w 但是最后一个chunkSize 却是达到了 600多w  最终导致最后一个check同步的时候失败了 并出现了这种错误 
![image](https://user-images.githubusercontent.com/26463863/150452958-d88b8b34-3cfa-4018-aa44-a7157325a02b.png)
这是什么原因呢？有没有办法解决？ @wuchong @xiaoyu19950911

**Reply by wuchong** [reply]
你这个表一直有在大量插入数据吗？

**Reply by wuchong** [reply]
chunks 是提前分好的，会不会分好 chunks 以后，到开始读最后一个 chunk 之间，插入了~600w 的数据？

**Reply by a120610114** [reply]
数据量也不是很大总共才1700多万，数据会有变更，但是也不会有太大 可能每天也就才几万的变更。最主要的是我是在晚上11点多开始同步的，这个时间段数据的变更是比较少的。

**Reply by a120610114** [reply]
是这样的，chunks一开始我看就是不对的190*5=950w 和实际数据就不符了呀
![image](https://user-images.githubusercontent.com/26463863/150454008-7aba35f2-85e9-465c-848d-724628a969e7.png)

**Reply by a120610114** [reply]
这种情况只出现了在这一张表中，其他表的chunks 又是对的，让我很疑惑

**Reply by xiaoyu19950911** [reply]
2.1版本是会对chunkSize进行动态调整的，但是理论上不应该分配不均匀，你可以下搜一下以下日志，看看对应的真实chunk划分规则
![image](https://user-images.githubusercontent.com/31729765/150457992-86538421-a6d2-4b97-8ad6-f4e6b171036e.png)

**Reply by a120610114** [reply]
看了一下代码，均匀分布的chunk 目前只支持数字型的主键，而我的主键是varchar 所以走的是不均匀分布的代码
![image](https://user-images.githubusercontent.com/26463863/150485130-b1740cdf-4363-4453-bff8-48a2e74ad5bf.png)


最终通过日志以及代码比对上找到了问题所在：
![image](https://user-images.githubusercontent.com/26463863/150484021-1fd84d26-83e9-4110-b3fa-1317dd5d442e.png)

![image](https://user-images.githubusercontent.com/26463863/150484087-91684ae9-95c0-4763-a5d4-348526b11912.png)

我们发现这个比较函数是有问题的，中间的id 居然比最后最大的一条id还大。就是这个导致最后一个chunk 分布不均匀。这个算是BUG吗？
还是mysql 中字符串的比较与java Comparable 比较的机制不同导致的？ @wuchong @xiaoyu19950911

**Reply by wuchong** [reply]
cc @leonardBang

**Reply by a120610114** [reply]
@wuchong @xiaoyu19950911 @leonardBang  经过排查 是因为mysql 中的字符比较是忽略大小写的，这是很多数据库的一种常见方式。 但是在java Comparable 的比较中却是不忽略大小写的。 按照ASCII 编码 小写字母就是比大写字母大。这就是导致chunks 分布不均匀的原因。 建议说提供一个参数来控制是否大小写敏感 或者就忽略大小写敏感是否更好呢？

**Reply by leonardBang** [reply]
> @wuchong @xiaoyu19950911 @leonardBang 经过排查 是因为mysql 中的字符比较是忽略大小写的，这是很多数据库的一种常见方式。 但是在java Comparable 的比较中却是不忽略大小写的。 按照ASCII 编码 小写字母就是比大写字母大。这就是导致chunks 分布不均匀的原因。 建议说提供一个参数来控制是否大小写敏感 或者就忽略大小写敏感是否更好呢？

@a120610114 非常感谢反馈和详细的排查，点赞！
这个比较应该和数据库保持一致，大多数数据库是忽略大小写的，那我们也应该忽略大小写，如果有的数据库是不忽略大小写的，那我们再去引入一个参数。有兴趣开个PR来修复吗？

**Reply by a120610114** [reply]
是的非常有兴趣

**Reply by gwqlovelinux** [reply]
scan.incremental.snapshot.enabled = false 
回退到老的版本，不使用2.1的特性，这样就快起来了，小表建议加这个参数
